### PR TITLE
feat: prevent serialisation of private component fields

### PR DESCRIPTION
**Pull request by BenjaminAmos**
### Contains

~~This pull request adds a new `ComponentTypeHandlerFactory` type specifcally for serialising components. It is mostly the same as `ObjectFieldMapTypeHandlerFactory` (from `TypeHandlerFactory`) but with an additional check for private fields. It might be possible to merge the two classes instead, this is just a first attempt to see how it behaves.~~

~~It is not possible to exclude private field serialiation from `ObjectFieldMapTypeHandlerFactory` completely since NUI depends on this behaviour extensively for UI widgets.~~

This pull request removes the ability to serialise private fields in `TypeHandlerLibrary`, apart from in a few special cases (mainly NUI). It changes the generic `ObjectFieldMapHandler` to use getter and setter methods where available instead.

### How to test

- Create a new game
- Play around for a bit (destroying blocks, loading chunks, etc.)
- Save the game
- Restart the game and load that save
- The world and surrounding entities should be similar to how you left it
- There should also not be any additional exceptions thrown due to these changes

### CODE REVIEW

**Review by soloturn — DISMISSED** [review]

**Comment by jdrueckert** [comment]
cross-post from Discord for the record:
I tested this branch with JS and ran into the following: https://pastebin.com/raw/rxsj7xWr
```
18:07:13.961 [parallel-1] ERROR reactor.core.publisher.Operators - Operator called default onErrorDropped
reactor.core.Exceptions$ErrorCallbackNotImplemented: java.lang.reflect.InaccessibleObjectException: Unable to make field private final java.lang.Class java.lang.Class.componentType accessible: module java.base does not "opens java.lang" to unnamed module @75eeccf5
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make field private final java.lang.Class java.lang.Class.componentType accessible: module java.base does not "opens java.lang" to unnamed module @75eeccf5
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
[...]

	at java.base/java.lang.reflect.Field.setAccessible(Field.java:172)
	at org.terasology.persistence.typeHandling.coreTypes.factories.ObjectFieldMapTypeHandlerFactory.lambda$getResolvedFields$1(ObjectFieldMapTypeHandlerFactory.java:80)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:318)
	at org.terasology.persistence.typeHandling.coreTypes.factories.ObjectFieldMapTypeHandlerFactory.getResolvedFields(ObjectFieldMapTypeHandlerFactory.java:68)
	at org.terasology.persistence.typeHandling.coreTypes.factories.ObjectFieldMapTypeHandlerFactory.create(ObjectFieldMapTypeHandlerFactory.java:40)
	at org.terasology.persistence.typeHandling.TypeHandlerLibrary.getTypeHandler(TypeHandlerLibrary.java:274)
```

what I did in JS: I gave myself a chest and some furnaces and sifting tables, placed them, put some stuff in them and then tried to exit to main menu (and it was a fresh TS workspace with JS recursed)

@BenjaminAmos responded:
> Yep, I knew about that issue. It's down to https://github.com/MovingBlocks/Terasology/blob/14937d24b851b300b3b8f569cfb86205a29626f2/engine/src/main/java/org/terasology/engine/logic/common/RetainComponentsComponent.java#L20. Technically it's a public field, so we try to serialise it. The actual serialiser then still looks at private fields inside of the public field value. It needs more thought.

**Review by jdrueckert — CHANGES_REQUESTED** [review]
We need to figure out how to deal with the `RetainComponentsComponent` issue before we can merge this.
Although I'm wondering why it shows up ... didn't I change all private component fields to public? :thinking:

File: subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/ObjectFieldMapTypeHandler.java
Comment (jdrueckert) [inline subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/ObjectFieldMapTypeHandler.java]: do we need something similar below for the setter?

File: subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/factories/ObjectFieldMapTypeHandlerFactory.java
Comment (jdrueckert) [inline subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/factories/ObjectFieldMapTypeHandlerFactory.java:93]: Why fluent API usage here?
I would expect `getName()` and `getTypeName()` to be O(1) so pretty much neglectable... also this is not on the happy path so it's not expected to be evaluated often...
In my test run, I got the warning only twice:
```
19:18:19.310 [main] WARN  o.t.p.t.c.f.ObjectFieldMapTypeHandlerFactory - Field exists#org.terasology.engine.entitySystem.entity.internal.PojoEntityRef will not be serialised. Terasology no longer supports serialising private fields.
19:18:19.396 [parallel-1] WARN  o.t.p.t.c.f.ObjectFieldMapTypeHandlerFactory - Field entityRef#org.terasology.engine.persistence.internal.DelayedEntityRef will not be serialised. Terasology no longer supports serialising private fields.
```

**Review by jdrueckert — DISMISSED** [review]
Tests out fine and looks good to me.
Only question: can we have some additional tests for the new behavior?

File: subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/factories/ObjectFieldMapTypeHandlerFactory.java
Comment (BenjaminAmos) [reply subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/factories/ObjectFieldMapTypeHandlerFactory.java:93]: I did not make that change. @soloturn pushed https://github.com/MovingBlocks/Terasology/pull/5208/commits/873fe5bf2594653225e4c525333d5dfc93e690eb to this branch, possibly as a well-intentioned fix for a newly-introduced checkstyle warning.

Comment (jdrueckert) [reply subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/factories/ObjectFieldMapTypeHandlerFactory.java:93]: okay, guess it might've been a PMD warning in that case.
I'd prefer suppressing the warning (suffix with `//NOPMD`) here due to the above-mentioned reasons.

File: subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/ObjectFieldMapTypeHandler.java
Comment (BenjaminAmos) [reply subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/ObjectFieldMapTypeHandler.java]: I believe that the setter is caught in the general catch block here:
https://github.com/MovingBlocks/Terasology/blob/ae83461fe1f6bca1b1c4706b976e49d38286066a/subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/ObjectFieldMapTypeHandler.java#L110-L112

I will add a more specific catch for that error to make it less ambiguous though.

**Comment by soloturn** [comment]
a lot better than before. with joshuasworld i get, and i seem recall that i had this before and it was my setup, some module missing isn't it?
```
19:26:59.664 [main] ERROR o.t.e.world.generation.WorldBuilder - Facet provider for class org.terasology.caves.CaveFacet is missing. It is required by class org.terasology.caves.CaveLocationProvider
19:26:59.664 [main] ERROR o.t.engine.core.modes.StateLoading - Error while loading org.terasology.engine.core.modes.loadProcesses.InitialiseWorldGenerator@28466003
java.lang.IllegalStateException: Missing facet provider
        at org.terasology.engine.world.generation.WorldBuilder.determineProviderChains(WorldBuilder.java:188)
        at org.terasology.engine.world.generation.WorldBuilder.build(WorldBuilder.java:95)
        at org.terasology.engine.world.generation.BaseFacetedWorldGenerator.getWorld(BaseFacetedWorldGenerator.java:83)
        at org.terasology.engine.world.generation.BaseFacetedWorldGenerator.initialize(BaseFacetedWorldGenerator.java:58)
        at org.terasology.engine.core.modes.loadProcesses.InitialiseWorldGenerator.step(InitialiseWorldGenerator.java:34)

```

File: subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/factories/ObjectFieldMapTypeHandlerFactory.java
Comment (soloturn) [reply subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/factories/ObjectFieldMapTypeHandlerFactory.java:93]: i am a little confused  - git blame says @BenjaminAmos changed that line. commit 873fe5b changed ComponentTypeHandlerFactory, different file, but the change looks the same? in case i really did this, feel free to undo. 2 warnings, one line too long, and one for the logger are there in the line before.

**Comment by jdrueckert** [comment]
@soloturn yes you faced that before: https://github.com/Terasology/JoshariasSurvival/issues/76
and I also noticed it sporadically happening (actually relatively often), but it's also happening with develop, so it seems unrelated to this change.
@BenjaminAmos IIRC you wanted to have a look at adding some test cases, right?
should we hold off merging until then or do you want to add them with a follow-up PR?

**Review by soloturn — APPROVED** [review]
not hold off merging. this improves a lot and finally saves stuff again.

**Pull request merged**
